svc-systems/plugin/client.go

```diff
@@ -0,0 +1,350 @@
+		}
+	}
+	return common.GeneralError(http.StatusNotFound, response.ResourceNotFound, "", []interface{}{"Chassis", uri}, nil)
+}
+
+func (m *multiTargetClient) Delete(uri string) response.RPC {
+	for _, target := range m.targets {
+		client := m.createClient(target)
+		resp := client.Delete(uri)
+		switch {
+		case resp.StatusCode == http.StatusNotFound:
+			continue
+		case is2xx(int(resp.StatusCode)):
+			return resp
+		case is4xx(int(resp.StatusCode)):
+			return resp
+		default:
+			log.Printf("execution of DELETE %s on %s plugin returned non 2xx status code; %v", uri, target.ID, resp.Body)
+		}
+	}
+	return common.GeneralError(http.StatusNotFound, response.ResourceNotFound, "", []interface{}{"Chassis", uri}, nil)
+}
+
+type Client interface {
+	Get(uri string, opts ...CallOption) response.RPC
+	Post(uri string, body *json.RawMessage) response.RPC
+	Patch(uri string, body *json.RawMessage) response.RPC
+	Delete(uri string) response.RPC
+}
+
+type client struct {
+	translator *uriTranslator
+	plugin     smodel.Plugin
+}
+
+func (c *client) Delete(uri string) response.RPC {
+	url := fmt.Sprintf("https://%s:%s%s", c.plugin.IP, c.plugin.Port, uri)
+	url = c.translator.toSouthbound(url)
+	resp, err := pmbhandle.ContactPlugin(url, http.MethodDelete, "", "", nil, map[string]string{
+		"UserName": c.plugin.Username,
+		"Password": string(c.plugin.Password),
+	})
+	return c.extractResp(resp, err)
+}
+
+func (c *client) Post(uri string, body *json.RawMessage) response.RPC {
+	url := fmt.Sprintf("https://%s:%s%s", c.plugin.IP, c.plugin.Port, uri)
+	url = c.translator.toSouthbound(url)
+	*body = json.RawMessage(c.translator.toSouthbound(string(*body)))
+	resp, err := pmbhandle.ContactPlugin(url, http.MethodPost, "", "", body, map[string]string{
+		"UserName": c.plugin.Username,
+		"Password": string(c.plugin.Password),
+	})
+	return c.extractResp(resp, err)
+}
+
+func (c *client) Patch(uri string, body *json.RawMessage) response.RPC {
+	url := fmt.Sprintf("https://%s:%s%s", c.plugin.IP, c.plugin.Port, uri)
+	url = c.translator.toSouthbound(url)
+	*body = json.RawMessage(c.translator.toSouthbound(string(*body)))
+	resp, err := pmbhandle.ContactPlugin(url, http.MethodPatch, "", "", body, map[string]string{
+		"UserName": c.plugin.Username,
+		"Password": string(c.plugin.Password),
+	})
+	return c.extractResp(resp, err)
+}
+
+func (c *client) Get(uri string, _ ...CallOption) response.RPC {
+	url := fmt.Sprintf("https://%s:%s%s", c.plugin.IP, c.plugin.Port, uri)
+	url = c.translator.toSouthbound(url)
+	resp, err := pmbhandle.ContactPlugin(url, http.MethodGet, "", "", nil, map[string]string{
+		"UserName": c.plugin.Username,
+		"Password": string(c.plugin.Password),
+	})
+
+	return c.extractResp(resp, err)
+}
+
+func (c *client) extractResp(httpResponse *http.Response, err error) response.RPC {
+	if err != nil {
+		return common.GeneralError(http.StatusInternalServerError, response.InternalError, err.Error(), nil, nil)
```

Comment:
Whenever  the request is made to the plugin   if the plugin is not reachable retry logic is added in which it will call /ODIM/v1/Status   with these two configuration variables `MaxRetryAttempt` https://github.com/ODIM-Project/ODIM/blob/development/lib-utilities/config/odimra_config.json#L87 and `RetryIntervalInMins` https://github.com/ODIM-Project/ODIM/blob/development/lib-utilities/config/odimra_config.json#L88, please add logic to handle the same

Reply:
Relying at minimum reasonable value obtained from equation:
`MaxRetryAttempt`  * `RetryIntervalInMins`  

gives you result of 1min which in case of synchronous GET query is far to long. Nobody wants to wait minute(s) for getting result of GET request.

What is a purpose of this mechanism? If necessary user can retry operation manually.  Isn't it good enough?

Reply:
I agree with Robert on this. We should fail and expect the NB client to retry.